pluginExamples/OEM/Worek.py

```diff
@@ -48,35 +30,174 @@
 activity = (sample mE - blank mE) / 10.6
 AChE = activity * 1.58e3
 activity = (sample mE - blank mE) * 1.58e3 / 10.6 
+
+=== PLUGINS IN ENLIGHTEN ===
+
+- get_intensity_from_wavelength
+- get_widget_from_name
+
 """
 
+def getY(x, domain, codomain):
+    """
+    given two arrays (domain, codomain) which describe a collection of points
+      domain: x0 x1 x2 x3 x4 ... xn
+    codomain: y0 y1 y2 y3 y4 ... yn
+
+    return the element from codomain such that the corresponding element from domain = x
+    ex: getY(4, [0,2,4,6,8], [1,3,5,7,9]) --> 5
+
+    if x is not in the domain, return the closest boundary element
+    """
+
+    for i in range(len(domain)-1):
+        if domain[i] <= x < domain[i+1]:
+            return codomain[i]
+
+    if domain[-1] >= x:
+        return codomain[-1]
+    
+    return codomain[0]
+
+def get_intensity_from_wavelength(wavelength, request):
+
+    wl_arr = request.settings.wavelengths
+    spectrum = request.processed_reading.processed
+
+    return getY(wavelength, wl_arr, spectrum)
+
+ChE_label = "ChE Activity"
+Hb_label = "Hb Content"
+
+ChE_Blank_slope_label = "ChE Blank Slope"
+ChE_Sample_slope_label = "ChE Sample Slope"
+
 class Worek(EnlightenPluginBase):
 
     def __init__(self):
         super().__init__()
+        self.clear_graph()
+
+    def start_recording(self):
+        self.clear_graph()
+        self._isrecording = True
+
+    def stop_recording(self):
+        self._isrecording = False
+
+    def clear_graph(self):
+        self._isrecording = False
+        self.startTime = time.time()
 
         self.sampleTimes = []
         self.ChEActivity = []
         self.HbContent = []
 
-        self.startTime = time.time()
+        self.slope_start = None
+        self.slope_end = None
+
+        self.ChEActivityY2 = None
+        self.ChEActivityY1 = None
+
+    def get_widget_from_name(self, name):
+        widget = None
+        for elem in self.enlighten_info.plugin_fields():
+            if elem.field_name == name:
+                widget = elem
+        return widget.field_widget
+
+    def graph_slope(self):
+        # self.get_widget_from_name("slope start") will be a QtWidgets.QDoubleSpinBox
+        # same thing for slope_end_txtbox
+
+        blank_start = self.get_widget_from_name("blank start").value()
```

Comment:
Do the calls to the widgets mean that this function has to run from the foreground thread?

Reply:
No, this is happening from the worker thread.

Reply:
We're able to read Qt widgets from other threads?  Is it just write access that is problematic?